src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs

```diff
@@ -154,12 +199,54 @@ private IEnumerable<TDocument> DeserializeDocuments(IEnumerable<RawBsonDocument>
 
             if (lastRawDocument != null)
             {
-                _changeStreamOperation.DocumentResumeToken = lastRawDocument["_id"].DeepClone().AsBsonDocument;
+                _documentResumeToken = lastRawDocument["_id"].DeepClone().AsBsonDocument;
             }
 
             return documents;
         }
 
+        private ResumeStartValue GetEffectiveResumeStartValues()
+        {
+            if (_postBatchResumeToken != null)
+            {
+                return new ResumeStartValue { ResumeAfter = _postBatchResumeToken };
+            }
+
+            if (_documentResumeToken != null)
+            {
+                return new ResumeStartValue { ResumeAfter = _documentResumeToken };
+            }
+
+            if (_startAfter != null)
+            {
+                return new ResumeStartValue { ResumeAfter = _startAfter };
+            }
+
+            if (_resumeAfter != null)
+            {
+                return new ResumeStartValue { ResumeAfter = _resumeAfter };
+            }
+
+            if (_startAtOperationTime != null || _initialOperationTime != null)
+            {
+                return new ResumeStartValue { StartAtOperationTime = _startAtOperationTime ?? _initialOperationTime };
+            }
+
+            return new ResumeStartValue { ResumeAfter = _resumeAfter, StartAfter = _startAfter, StartAtOperationTime = _startAtOperationTime };
+        }
+
+        private BsonDocument GetPostBatchResumeTokenOrThrow()
```

Comment:
I don't think this method is worth it. Just cast _cursor directly to ICursorBatchInfo when needed.

I don't see any reason we need a custom error message for an InvalidCastException that will never be thrown.

Reply:
Done